### ISSUE TITLE
Add unique test cases to SimpleClassifierList

### ISSUE THREAD

**Issue by skekre98** [skekre98/simple_learn#30]
**Description of feature/enhancement**
Currently `SimpleClassifierList` is only being tested against two different datasets, wine and iris. It would be a good idea to verify this class against more datasets.  

**Description of implementation**
Add the test cases to the `TestSimpleClassifierList` class:
https://github.com/skekre98/simple_learn/blob/953150faecf12951407e2aed3418fd139b0ab9ed/tests/classifiers/simple_classifier_list_tests.py#L33

Recommended to use pre-wrangled datasets such as those available in `sklearn`.

**Why this is a good feature/enhancement**
Will allow us to verify the class against a wider variety of datasets.

**Additional context**
In the future might be nice to create in house datasets for testing.

**Comment by eranns** [comment]
Hi

I would like to help you by contributing a few tests. 
I noticed a bug in `def_test_iris` : 

```
def test_iris(self):

        iris = datasets.load_wine()
```

you load wine instead of iris dataset if I got it right.

**Comment by skekre98** [comment]
Hi @eranns, you are absolutely right. This is indeed a bug, that line should read `iris = datasets.load_iris()`.  Thanks for catching that! I'll go ahead and assign the issue over to you 😄 . Can you also fix the iris dataset in your PR as well?

**Comment by eranns** [comment]
Yes sure

**Comment by eranns** [comment]
Hey @skekre98.

 While creating tests for more datasets from `sklearn` I noticed some models used by `simpleClassifierList` does not support `multilabel classification` problems or `negative values in  the data`.

is it by design ? 

regarding the negative values, It happens only with `ComplementNB` classifier because this model assume for only positive data.

Models support multilabel in scikit-learn documented [here](https://scikit-learn.org/stable/modules/multiclass.html)

**Comment by skekre98** [comment]
> Hey @skekre98.
> 
> While creating tests for more datasets from `sklearn` I noticed some models used by `simpleClassifierList` does not support `multilabel classification` problems or `negative values in the data`.
> 
> is it by design ?
> 
> regarding the negative values, It happens only with `ComplementNB` classifier because this model assume for only positive data.
> 
> Models support multilabel in scikit-learn documented [here](https://scikit-learn.org/stable/modules/multiclass.html)

Hi @eranns, yes the lack of _multilabel_ support is currently by design. I consider _multilabel_ and _multiclass_ two different data modeling problems. In the future, I think we should have a class attribute called `label_type` which will trigger two different model selection strategies(perhaps two different parameter grids). For now, I think it will be fine to stick with _multiclass_ data. WDYT?

I was unaware of issues with negative values, but if it is only causing problems with a singular model algorithm do you think it would be a good idea if we just handled it with a `try/except` block? I am open to other implementations as well.

**Comment by eranns** [comment]
> which will trigger two different model selection strategies(perhaps two different parameter grids). For now, I think it will be fine to stick with multiclass data. WDYT?

sounds good.

> I was unaware of issues with negative values, but if it is only causing problems with a singular model algorithm do you think it would be a good idea if we just handled it with a try/except block? I am open to other implementations as well.

I'm not a fan of try/catch but otherwise this model would need to be removed. I think it depends if this model is important to the list of models - I haven't use this model in the past.

If a try/catch is the solution it should be in the `simpleclassifierList` fit, probably in another issue as bug related.

**Comment by skekre98** [comment]
> I'm not a fan of try/catch but otherwise this model would need to be removed. I think it depends if this model is important to the list of models - I haven't use this model in the past.

Essentially what we don't want is the entire `fit` function failing for `SimpleClassifierList` if only one model algorithm is throwing an error. Let's perhaps handle this in a separate issue. What I had in mind was keeping track of the models that are being skipped and logging this at the end to let the user know. 

Ex.
```python
for clf_class in classifiers:
    try:
        gridsearch.fit(clf_class)
    except ModelError:
        fail_models.append(clf_class.name)

log(failed_models)
```

I agree this isn't ideal since it would be best not to completely skip algorithms. Perhaps we could even have a failure threshold so if a certain number of models failed we can notify the user. Let me know if you have anything better in mind.

> If a try/catch is the solution it should be in the `simpleclassifierList` fit, probably in another issue as bug related.

Agreed

**Comment by eranns** [comment]
> What I had in mind was keeping track of the models that are being skipped and logging this at the end to let the user know.

That seems reasonable If it's just one two out of many models

>  agree this isn't ideal since it would be best not to completely skip algorithms. Perhaps we could even have a failure threshold so if a certain number of models failed we can notify the user. Let me know if you have anything better in mind.

I think your idea is good enough.  
saving failed models can also be useful for later -   In a way this library is making life easier for people without machine learning knowledge but this list can also help more advanced users see which models failed and perhaps understand better their data.



Regarding the tests  - I'm using sklearn.datasets and managed to add two more small datasets like Iris  but when I test on real world datasets from sklearn.datasets I was unable to complete the tests because even after an hour it doesn't seem to finish grid search on my computer (using only dual core :( )
I don't want to add tests I haven't  been able to complete myself - tell me what do you think, maybe other people with better CPU can contribute and add it later.

**Comment by skekre98** [comment]
> I think your idea is good enough. Saving failed models can also be useful for later - In a way this library is making life easier for people without machine learning knowledge but this list can also help more advanced users see which models failed and perhaps understand better their data.

Cool, I'll go ahead and create another issue for logging failed models. If it turns out more than a couple models are failing, I'll update the issue down the road.

> Regarding the tests - I'm using sklearn.datasets and managed to add two more small datasets like Iris but when I test on real world datasets from sklearn.datasets I was unable to complete the tests because even after an hour it doesn't seem to finish grid search on my computer (using only dual core :( )

The repository is integrated with Github Actions, so when you open a PR the tests will be triggered. If the tests fail, we can see where they failed and figure out how to move forward. We can discuss it more on the PR thread :)

**Comment by skekre98** [comment]
Created logging issue here: #36

**Comment by skekre98** [comment]
Merged in #37
